src/hist/core.py

```diff
@@ -45,24 +45,32 @@ def __call__(self, *args, **kwargs):
 
 @hist.utils.set_family(hist.utils.HIST_FAMILY)
 class BaseHist(bh.Histogram):
+    __slots__ = ("_ax", "_storage_proxy")
+
     def __init__(self, *args, **kwargs):
         """
             Initialize BaseHist object. Axis params can contain the names.
         """
+        self._ax: List[Any] = []
+        self._hist: Any = None
+        self._storage_proxy: Any = None
+
         if len(args):
-            super().__init__(*args, **kwargs)
-            self.names: dict = dict()
-            for ax in self.axes:
-                if ax.name and ax.name in self.names:
-                    raise KeyError(
-                        f"{self.__class__.__name__} instance cannot contain axes with duplicated names."
-                    )
-                else:
-                    self.names[ax.name] = True
-        else:
-            self._ax = []
             self._hist = None
-            self._storage_proxy = None
+            self._ax = []
+            super().__init__(*args, **kwargs)
+            valid_names = [ax.name for ax in self.axes if ax.name]
+            if len(valid_names) != len(set(valid_names)):
+                raise KeyError(
+                    f"{self.__class__.__name__} instance cannot contain axes with duplicated names"
+                )
+            for i, ax in enumerate(self.axes):
```

Comment:
This adds default titles if both name and title are unset. Actually, now that I think about it, we probably only need `if ax.title`, since we now have the title draw from the name if unset...